src/Tools/ExternalAccess/VisualDiagnostics/Internal/VisualDiagnosticsServiceFactory.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.BrokeredServices;
+using Microsoft.CodeAnalysis.ExternalAccess.VisualDiagnostics.Contracts;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.ServiceHub.Framework;
+using Roslyn.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VisualDiagnostics;
+
+/// <summary>
+/// LSP Service responsible for loading IVisualDiagnosticsLanguageService workspace service and delegate the broker service to the workspace service,
+/// and handling MAUI XAML/C#/CSS/Razor Hot Reload support
+/// </summary>
+[Export(typeof(IOnServiceBrokerInitialized))]
+[ExportCSharpVisualBasicLspServiceFactory(typeof(OnInitializedService)), Shared]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+[method: ImportingConstructor]
+internal sealed class VisualDiagnosticsServiceFactory(
+    LspWorkspaceRegistrationService lspWorkspaceRegistrationService) : ILspServiceFactory, IOnServiceBrokerInitialized
+{
+    private readonly LspWorkspaceRegistrationService _lspWorkspaceRegistrationService = lspWorkspaceRegistrationService;
+    private readonly Lazy<OnInitializedService> _OnInitializedService = new Lazy<OnInitializedService>(() => new OnInitializedService(lspWorkspaceRegistrationService));
+
+    public ILspService CreateILspService(LspServices lspServices, WellKnownLspServerKinds serverKind)
+    {
+        return _OnInitializedService.Value;
+    }
+
+    public void OnServiceBrokerInitialized(IServiceBroker serviceBroker)
+    {
+        _OnInitializedService.Value.OnServiceBrokerInitialized(serviceBroker);
+    }
+
+    private class OnInitializedService : ILspService, IOnInitialized, IOnServiceBrokerInitialized, IDisposable
+    {
+        private readonly LspWorkspaceRegistrationService _lspWorkspaceRegistrationService;
+        private IVisualDiagnosticsLanguageService? _visualDiagnosticsLanguageService;
+        private CancellationToken _cancellationToken;
+        private static readonly TaskCompletionSource<bool> _taskCompletionSource = new TaskCompletionSource<bool>();
```

Comment:
In general, this kind of TCS use can be problematic when entrypoints into the class cannot complete without this 'black box' also being completed. It *can* lead to deadlocks. 
As you have it here, I don't see such a risk, but I try to avoid the pattern anyway because these things tend to grow organically until it's huge and expensive to re-architect.
If you're not running this in devenv.exe, it's probably much less of an issue.

Reply:
Because two LSP services are initializing in parallel, I cannot right now of anything simpler than using a TCS.  By design, the entry point into this class is likely to always complete because the LSP service is guaranteed to be initialized. The broker service on the other is not guaranteed to be initially and thus the call to OnServiceBrokerInitialized will never be made.  But the TCS is guaranteed to initialized in his normal flow.  Correct this is not going be run inside devenv.